drivers/lora/sx1276.c

```diff
@@ -0,0 +1,717 @@
+
+	/* Set DIO0 to Tx done */
+	ret = sx1276_read(dev, SX1276_REG_DIO_MAPPING1, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read DIO_MAPPING1");
+		return -EIO;
+	}
+
+	regval &= ~REG_DIO_MAPPING1_DIO0_MASK;
+	regval |= 0x1 << 6;
+	ret = sx1276_write(dev, SX1276_REG_DIO_MAPPING1, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write DIO_MAPPING1");
+		return -EIO;
+	}
+
+	ret = sx1276_read(dev, SX1276_REG_OPMODE, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read OPMODE");
+		return -EIO;
+	}
+
+	/* FIFO cannot be accessed in sleep mode */
+	if (regval & SX1276_OPMODE_MODE_SLEEP) {
+		regval &= ~SX1276_OPMODE_MODE_MASK;
+		regval |= SX1276_OPMODE_MODE_STDBY;
+	}
+
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	/* Make use of full FIFO for Tx */
+	ret = sx1276_write(dev, SX1276_REG_FIFO_TX_BASE_ADDR, 0x00);
+	if (ret < 0) {
+		LOG_ERR("Unable to write FIFO Tx base addr");
+		return -EIO;
+	}
+
+	ret = sx1276_write(dev, SX1276_REG_FIFO_ADDR_PTR, 0x00);
+	if (ret < 0) {
+		LOG_ERR("Unable to write FIFO addr pointer");
+		return -EIO;
+	}
+
+	ret = sx1276_write(dev, SX1276_REG_PAYLOAD_LENGTH, data_len);
+	if (ret < 0) {
+		LOG_ERR("Unable to write payload length");
+		return -EIO;
+	}
+
+	ret = sx1276_fifo_write(dev, data, data_len);
+	if (ret < 0)
+		return ret;
+
+	ret = sx1276_read(dev, SX1276_REG_OPMODE, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read OPMODE");
+		return -EIO;
+	}
+
+	regval &= ~SX1276_OPMODE_MODE_MASK;
+	regval |= SX1276_OPMODE_MODE_TX;
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int sx1276_lora_config(struct device *dev,
+			      struct lora_modem_config *config)
+{
+	unsigned long long freq_rf;
+	int ret;
+	u8_t regval, pa_config;
+
+	/* Set sleep mode */
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, SX1276_OPMODE_MODE_SLEEP);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	/* Set LoRa mode */
+	ret = sx1276_read(dev, SX1276_REG_OPMODE, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read OPMODE");
+		return -EIO;
+	}
+
+	regval |= SX1276_OPMODE_LONG_RANGE_MODE;
+	/* Set low frequency mode if required */
+	if (config->frequency < 525000000)
+		regval |= SX1276_OPMODE_LOW_FREQUENCY_MODE_ON;
+
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	/* Set standby mode */
+	ret = sx1276_read(dev, SX1276_REG_OPMODE, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read OPMODE");
+		return -EIO;
+	}
+
+	regval &= ~SX1276_OPMODE_MODE_MASK;
+	regval |= SX1276_OPMODE_MODE_STDBY;
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	k_sleep(200);
+
+	/* Set frequency */
+	freq_rf = config->frequency;
+	freq_rf *= (1 << 19);
```

Comment:
You can safely ignore this comment, but I would prefer:
```C
freq_rf = ((u64_t)config->frequency << 19) / CLOCK_FREQ;
```
I think it's more clear, but maybe other people think differently :-)